framework/mutex/Mutex.php

```diff
@@ -97,6 +103,40 @@ public function release($name)
         return false;
     }
 
+    /**
+     * Executes callback with mutex synchronization.
+     *
+     * @param string $name of the lock to be acquired. Must be unique.
+     * @param int $timeout time (in seconds) to wait for lock to be released.
+     * @param callable $callback a valid PHP callback that performs the job. Accepts mutex instance as parameter.
+     * @param bool $throw whether to throw an exception when the lock is not acquired.
+     * @return mixed result of callback function, or null when the lock is not acquired.
+     * @throws \Exception|\Throwable if there is any exception thrown from the callback.
+     * @throws SyncException when the lock is not acquired.
+     * @since (unknown yet)
+     */
+    public function sync($name, $timeout, callable $callback, $throw = true)
+    {
+        if ($this->acquire($name, $timeout)) {
+            try {
+                $result = call_user_func($callback, $this);
+            } catch (\Exception $e) {
+                $this->release($name);
+                throw $e;
+            } catch (\Throwable $e) {
+                $this->release($name);
+                throw $e;
+            }
+            // TODO: use "finally" keyword when miminum required PHP version is >= 5.5
```

Comment:
This PR contains a new feature, so will go to 2.1 branch. You can already use `finally` keyword

Reply:
Done